backend/corpora/dataset_processing/remaster_cxg.py

```diff
@@ -27,48 +32,151 @@ def process(dataset_id: str, cellxgene_bucket: str, prefix=None, dry_run=True):
+        else:
+            logger.info("Dataset was already upgraded")
 
     # Cleanup
     logger.info("Cleaning up local files")
+    _try_to_delete(f"{local_path}/X_old")
+    _try_to_delete(f"{local_path}/X_newr")
+    _try_to_delete(f"{local_path}/X_newc")
+    _try_to_delete(f"{local_path}/old_obs")
+    _try_to_delete(f"{local_path}/new_obs")
+    _try_to_delete(f"{local_path}/cxg_group_metadata")
+
+
+def _try_to_delete(path):
     import shutil
 
-    shutil.rmtree(f"{local_path}/X_old")
-    shutil.rmtree(f"{local_path}/X_new")
+    try:
+        shutil.rmtree(path)
+    except FileNotFoundError:
+        pass
+
 
+def increment_version(cxg):
+    """
+    Increments the version number of the CXG
+    :param cxg: the path to the CXG
+    """
+    with tiledb.open(f"{cxg}/cxg_group_metadata", "w") as X:
+        X.meta["cxg_version"] = CxgConstants.CXG_VERSION
 
-def compute(**kwargs):
+
+def evolve_obs(cxg, array_name="old_obs"):
+    """
+    Creates a new obs array with the new schema
+    :param X: the old obs tiledb array
     """
-    Computes the evolved cxg from source_array and saves it to the target_array destination
+    with tiledb.open(f"{cxg}/{array_name}", "r") as X:
+        # load cxg schema
+        schema = json.loads(X.meta["cxg_schema"])
+
+        # get all data in the old obs array
+        data = X.query().multi_index[:]
+
+        # adjust the schema to expect codes where applicable and store code-to-value mapping
+        # dictionary in type_hint["categories"]
+        tdb_attrs = []
+        new_data = {}
+        for attr in X.schema:
+            a = attr.name
+            type_hint = schema.get(a, {})
+            if "categories" in type_hint and len(type_hint.get("categories", [])) > 0.75 * X.shape[0]:
+                schema[a]["type"] = "string"
+                del schema[a]["categories"]
+                tdb_attrs.append(attr)
+                new_data[a] = data[a]
+            elif "categories" in type_hint:
+                cat = pd.Categorical(data[a])
+                codes = cat.codes
+                new_data[a] = codes
+                categories = cat.categories
+                schema[a]["categories"] = list(categories)
+
+                dtype = str(cat.codes.dtype)
+                tdb_attrs.append(tiledb.Attr(name=a, dtype=dtype, filters=attr.filters))
+            else:
+                tdb_attrs.append(attr)
+                new_data[a] = data[a]
+
+        new_schema = tiledb.ArraySchema(domain=X.schema.domain, attrs=tdb_attrs)
+        tiledb.Array.create(
+            f"{cxg}/new_obs",
+            new_schema,
+        )
+        with tiledb.open(f"{cxg}/new_obs", "w") as new_X:
+            new_X[:] = new_data
+            new_X.meta["cxg_schema"] = json.dumps(schema)
+        tiledb.consolidate(f"{cxg}/new_obs")
+
+
+def evolve_X(**kwargs):
+    """
+    Computes the evolved cxg from the sparse source_array and saves it to the target_array destination
```

Comment:
nit: mind adding the return type to the docstring + explain it briefly in the description?